### PR TITLE
Get field

### PR DESCRIPTION
**Pull request by rikner**
add `GetField()` to JNI, which takes a field name and and object and returns the value of that field of that object (if exists). Wraps `GetObjectClass()`, `GetFieldID` and `Get<Type>Field` of `JNI`.

### CODE REVIEW

File: Sources/JNI/JNIFields.swift [hunk 1]
```diff
@@ -17,4 +17,13 @@ public extension JNI {
         }
         return try T.fromStaticField(of: javaClass, id: fieldID)
     }
+
+    public func GetField<T: JavaParameterConvertible>(_ fieldName: String, from javaObject: JavaObject) throws -> T {
+        let env = self._env
+        guard let javaClass = GetObjectClass(obj: javaObject),
+              let fieldID = env.pointee.pointee.GetFieldID(env, javaClass, fieldName, T.asJNIParameterString) else {
```
Comment (cshg) [inline Sources/JNI/JNIFields.swift]: würde zur besseren Lesbarkeit das `else` auf die nächste Zeile einrücken

File: Sources/JNI/JNIObjects.swift [hunk 2]
```diff
@@ -1,22 +1,24 @@
...
+struct ConstructorError: Error {}
 
 /// Designed to simplify calling a constructor and methods on a JavaClass
 /// Subclass this and add the methods appropriate to the object you are constructing.
 open class JNIObject {
     public let javaClass: JavaClass
     public let instance: JavaObject
 
-    public init?(_ className: String, arguments: [JavaParameterConvertible] = []) {
+    public init(_ className: String, arguments: [JavaParameterConvertible] = []) throws {
```
Comment (cshg) [inline Sources/JNI/JNIObjects.swift:12]: :+1: finde error thrown nicer als nil

File: Sources/JNI/JavaParameterConvertible.swift [hunk 3]
```diff
@@ -35,6 +36,10 @@ extension Bool: JavaParameterConvertible {
     public static func fromStaticMethod(calling methodID: JavaMethodID, on javaClass: JavaClass, args: [JavaParameter]) throws -> Bool {
         return try jni.CallStaticBooleanMethod(javaClass: javaClass, method: methodID, parameters: args) == JNI_TRUE
     }
+
+    public static func fromField(of javaObject: JavaObject, id: jfieldID) throws -> Bool {
+        return jni.GetBooleanField(of: javaObject, id: id) == 1
```
Comment (cshg) [inline Sources/JNI/JavaParameterConvertible.swift]: returned `GetBooleanField` 1 / 0 statt true / false? können wir `jboolean` nicht direkt als Bool anstatt UInt8 definieren?

**Review by cshg — APPROVED** [review]
Sieht nice aus, vor allem die neuen Error throws :+1:
Habe nicht alles getestet, hauptsächlich das neue throwen beim init und in GetMethodID. Wäre generell nice in Zukunft noch Tests hier zu haben.
Ansonsten nur paar kleine Comments / Fragen.

File: Sources/JNI/JavaParameterConvertible.swift (hunk 3 above)
Comment (ephemer) [reply Sources/JNI/JavaParameterConvertible.swift]: `== JNI_TRUE` ist mMn die klarste Variante

jboolean wird von der JNI statt von uns definiert. Deswegen wandeln wir die Werte immer mit dem Vergleich, oben, um.

File: Sources/JNI/JNIFields.swift (hunk 1 above)
Comment (ephemer) [reply Sources/JNI/JNIFields.swift]: GetFieldID should also throw if it's checked via try checkAndThrow(), that way we should be able to avoid making our own error types

File: Sources/JNI/JNIObjects.swift [hunk 4]
```diff
@@ -1,22 +1,24 @@
...
         let className = className.replacingFullstopsWithSlashes()
 
         guard
             let javaClassLocalRef = jni.FindClass(name: className),
             let javaClass = jni.NewGlobalRef(javaClassLocalRef)
         else {
             assertionFailure("Couldn't find class named \(className)")
-            return nil
+            throw ClassNotFound()
```
Comment (ephemer) [inline Sources/JNI/JNIObjects.swift]: FindClass should also throw I think, then we can remove this error

File: Sources/JNI/JavaParameterConvertible.swift [hunk 5]
```diff
@@ -58,6 +63,10 @@ extension Int: JavaParameterConvertible {
         let result = try jni.CallStaticIntMethod(methodID, on: javaClass, parameters: args)
         return Int(result)
     }
+
+    public static func fromField(of javaObject: JavaObject, id: jfieldID) throws -> Int {
```
Comment (ephemer) [inline Sources/JNI/JavaParameterConvertible.swift]: We should be careful to keep the order of parameters consistent. My suggestion would be:

`public static func fromField(_ fieldID: JavaFieldID, on javaObejct: JavaObject) throws -> Int`

File: Sources/JNI/JavaParameterConvertible.swift [hunk 6]
```diff
@@ -104,4 +117,11 @@ extension String: JavaParameterConvertible {
         let jObject = try jni.CallStaticObjectMethod(methodID, on: javaClass, parameters: args)
         return jni.GetString(from: jObject)
     }
+
+    public static func fromField(of javaObject: JavaObject, id: jfieldID) throws -> String {
+        guard let javaStringObject = jni.GetObjectField(of: javaObject, id: id) else {
+            throw FieldIDNotFound()
```
Comment (ephemer) [inline Sources/JNI/JavaParameterConvertible.swift]: This error isn't entirely accurate, we could have an error here for other reasons. GetObject field should just throw, and this function should call it with try

File: Sources/JNI/SwiftJNI.swift [hunk 7]
```diff
@@ -56,10 +56,12 @@ extension JNI {
...
         try checkAndThrowOnJNIError()
 
-        let result = _env.pointee.pointee.GetMethodID(_env, objectClass!, methodName, methodSignature)
-        try checkAndThrowOnJNIError()
+        guard let result = _env.pointee.pointee.GetMethodID(_env, objectClass!, methodName, methodSignature) else {
+            throw InvalidMethodID()
+        }
 
-        return result!
+        try checkAndThrowOnJNIError()
```
Comment (ephemer) [inline Sources/JNI/SwiftJNI.swift]: Pretty sure the previous version was more accurate / safer because it'll check and clear any JNI error that occurs instead of making up our own and leaving any JNI error on the stack

File: Sources/JNI/SwiftJNI.swift [hunk 8]
```diff
@@ -162,6 +164,28 @@ extension JNI {
 
     // MARK: Fields
 
+    public func GetBooleanField(of javaObject: JavaObject, id: jfieldID) -> JavaBoolean {
+        let _env = self._env
+        return _env.pointee.pointee.GetBooleanField(_env, javaObject, id)
```
Comment (ephemer) [inline Sources/JNI/SwiftJNI.swift]: These should all throw and `try checkAndThrowOnJNIError()` before returning

File: Sources/JNI/SwiftJNI.swift [hunk 9]
```diff
@@ -215,7 +239,7 @@ extension JNI {
 
         var result = [JavaByte](repeating: 0, count: count)
         _env.pointee.pointee.GetByteArrayRegion(_env, array, jsize(startIndex), jsize(count), &result)
-        return result.map { UInt8(bitPattern: $0) } // may not have the same numeric value, but
+        return result.map { UInt8(bitPattern: $0) } // numeric value may change here because of the bitPattern constructor
```
Comment (ephemer) [inline Sources/JNI/SwiftJNI.swift]: bitPattern is very likely to be the wrong constructor here. What are we trying to achieve?

Comment (ephemer) [reply Sources/JNI/SwiftJNI.swift]: I thought JavaByte was just a typealias for UInt8, sure we need to make a copy of the array here?

File: Sources/JNI/SwiftJNI.swift [hunk 10]
```diff
@@ -284,6 +308,17 @@ extension JNI {
...
         return strings
     }
+
+    public func GetObjectArrayElement(in array: JavaObjectArray, at index: Int) throws -> JavaObject {
+        let _env = self._env
+        let count = jni.GetLength(array)
+        if (index >= count) {
+            throw JNIError()
+        }
+        let jObj = _env.pointee.pointee.GetObjectArrayElement(_env, array, jsize(index))!
```
Comment (ephemer) [inline Sources/JNI/SwiftJNI.swift]: This should not be force unwrapped with `!`. Instead, we should leave it optional, check for the JNI error (as below) and return a force-unwrapped jObj in that line

File: Sources/JNI/SwiftJNI.swift [hunk 11]
```diff
@@ -284,6 +308,17 @@ extension JNI {
 
         return strings
     }
+
+    public func GetObjectArrayElement(in array: JavaObjectArray, at index: Int) throws -> JavaObject {
```
Comment (ephemer) [inline Sources/JNI/SwiftJNI.swift:321]: I wonder whether we can make swift wrapper types for typed arrays

struct JavaArray<T: JavaType> {
  subscript(index: Int) -> T
}

Would leave this for another PR though

File: Sources/JNI/JavaParameterConvertible.swift [hunk 12]
```diff
@@ -15,6 +15,7 @@ public protocol JavaParameterConvertible {
     static func fromMethod(calling methodID: JavaMethodID, on object: JavaObject, args: [JavaParameter]) throws -> Self
     static func fromStaticMethod(calling methodID: JavaMethodID, on javaClass: JavaClass, args: [JavaParameter]) throws -> Self
     static func fromStaticField(of javaClass: JavaClass, id: jfieldID) throws -> Self
+    static func fromField(of javaObject: JavaObject, id: jfieldID) throws -> Self
```
Comment (ephemer) [inline Sources/JNI/JavaParameterConvertible.swift]: We should rename jfieldID to JavaFieldID in jni.h

File: Sources/JNI/JavaParameterConvertible.swift (hunk 5 above)
Comment (ephemer) [reply Sources/JNI/JavaParameterConvertible.swift]: Same for staticField

**Review by ephemer — CHANGES_REQUESTED** [review]

**Comment by ephemer** [comment]
I think in general we have the issue of sometimes throwing and sometimes returning optionals. I think throwing is more idiomatic JNI / swift, but we need to be consistent with that.

Also we should be careful to keep consistency within our protocols (eg argument order), not sure what the best way forward with that is.

File: Sources/JNI/SwiftJNI.swift (hunk 9 above)
Comment (rikner) [reply Sources/JNI/SwiftJNI.swift]: `JavaByte` is a typealias to `Int8`. So in order to make sure the program does not crash during the conversion from `Int8` to `UInt8`, we use  the `bitPattern` constructor. This way positive `Int8`s remain the same during conversion, a negative `Int8` will be mapped to a value between 128 and 255.

**Comment by rikner** [comment]
@ephemer I think I've done all the requested changes. Thanks for the helpful feedback!
